torch/utils/bundled_inputs.py

```diff
@@ -0,0 +1,148 @@
+        This function will attempt to optimize arguments so that (e.g.)
+        arguments like `torch.zeros(1000)` will be represented compactly.
+        Only top-level arguments will be optimized.
+        Tensors in lists or tuples will not.
+    """
+    if not isinstance(model, torch.jit.ScriptModule):
+        raise Exception("Only ScriptModule is supported.")
+
+    forward_arg_types = [node.type() for node in list(model.graph.param_node().outputs())[1:]]
+    deflated_inputs_type = torch._C.ListType(torch._C.TupleType(forward_arg_types))
+    inflated_inputs_type = torch._C.OptionalType(deflated_inputs_type)
+    model._c._register_attribute("_bundled_inputs_deflated", deflated_inputs_type, [])
+    model._c._register_attribute("_bundled_inputs_inflated", inflated_inputs_type, None)
+
+    if hasattr(model, "_generate_bundled_inputs"):
+        if inputs is not None:
+            raise Exception(
+                "inputs is not None, but _generate_bundled_inputs is already defined")
+        # Model author already defined _generate_bundled_inputs.
+    elif inputs is None:
+        raise Exception(
+            "inputs must be specified if _generate_bundled_inputs is not already defined")
+    else:
+        # Iterate over the inputs and args in each input.
+        # Accumulate `deflated_inputs` as (possibly) compressed values
+        # and `parts` to be joined into the expression that unpacks them.
+        deflated_inputs = []
+        parts = []
+        for inp_idx, args in enumerate(inputs):
+            deflated_args = []
+            parts.append("(")
+            for arg_idx, arg in enumerate(args):
+                deflated, inflater = _inflate_expr(arg, f"deflated[{inp_idx}][{arg_idx}]")
+                deflated_args.append(deflated)
+                parts.append(f"    {inflater},")
+            deflated_inputs.append(tuple(deflated_args))
+            parts.append("),")
+        parts.append("")
+        expr = "\n".join(parts)
+        # Back-channel return this expr for debugging.
+        if _receive_inflate_expr is not None:
+            _receive_inflate_expr.append(expr)
+        model._bundled_inputs_deflated = deflated_inputs
+        definition = textwrap.dedent("""
+            def _generate_bundled_inputs(self):
+                deflated = self._bundled_inputs_deflated
+                return [
+            {}
+                ]
+            """).format(expr)
+        model.define(definition)
+
+    # Define get_all_bundled_inputs that caches the generated inputs.
+    model.define(textwrap.dedent("""
+        def get_all_bundled_inputs(self):
+            if self._bundled_inputs_inflated is None:
+                self._bundled_inputs_inflated = self._generate_bundled_inputs()
+            all_inputs = self._bundled_inputs_inflated
+            assert all_inputs is not None
+            return all_inputs
+        """))
+
+    # Define some helper methods.
```

Comment:
My impression is that here is more code generation here than necessary. Couldn't this functionality be implemented as a normal TorchScript model, and then converted use torch.jit.script?

Reply:
I don't see how, but maybe I'm missing something.  I want this to be available as a utility that can be run after training, regardless of what the top-level module is.  Since I don't want to wrap the top-level module (see commit message), this was the only way I thought of to add these helpers.